File: algorithms/cpp/houseRobberII/houseRobberII.cpp
// Source: https://leetcode.com/problems/house-robber-ii/
// Author: Miao Zhang
// Date:   2021-01-26

class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.empty()) return 0;
        int n = nums.size();
        if (n == 1) return nums[0];
        if (n == 2) return max(nums[0], nums[1]);
        return max(robTotal(nums, 0, n - 2), robTotal(nums, 1, n - 1));
    }
    
    int robTotal(vector<int>& nums, int i, int j) {
        if (i + 1 >= j) return max(nums[i], nums[j]);
        vector<int> dp(j - i + 1, 0);
        dp[0] = nums[i];
        dp[1] = max(nums[i], nums[i + 1]);
        for (int k = i + 2; k <= j; k++) {
            dp[k - i] = max(dp[k - 1 - i], dp[k - 2 - i] + nums[k]);
        }
        return dp[j - i];
    }
};

File: algorithms/python/houseRobberII/houseRobberII.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/house-robber-ii/
# Author: Miao Zhang
# Date:   2021-01-26

class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        n = len(nums)
        if n == 1: return nums[0]
        if n == 2: return max(nums[0], nums[1])
        return max(self.robTotal(nums[0: n - 1]), self.robTotal(nums[1: n]))
    
    def robTotal(self, nums: List[int]) -> int:
        if not nums:
            return 0
        n = len(nums)
        if n == 1: return nums[0]
        if n == 2: return max(nums[0], nums[1])
        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, n):
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])
        return dp[n - 1]
